### ISSUE TITLE
Clang PCH + FastBuild Cache without -fno-pch-timestamp?

### ISSUE THREAD

**Issue by missmah** [fastbuild/fastbuild#346]
This just came up today, and I'm highly skeptical that we aren't doing something wrong on our end (we must be!...). But, it does look like this might be a generic issue that somehow nobody else has hit but us... so I'm going to ask in the meanwhile before we hopefully come to some solution ourselves...

Are there any issues with Clang `PCH`s and `FastBuild -cache` without `-fno-pch-timestamp`?

Logically, it does seem like Clang's current default behavior (to check timestamps of PCH files against the headers they `#include`) would be incompatible with `FastBuild`'s caching mechanism, which will cache PCHs even if their timestamps differ, based upon their preprocessed output matching the current version on disk's preprocessed output.

Okay, so here's the repro:
0. Build via `FastBuild` with `-cache` using a `Clang` compiler
1. `Touch` a central header that's included in a given `PCH` (don't modify the file, just `touch` the timestamp).
2. Modify (actually change) a `.cpp` file which depends upon this header, via a `PCH` which should be rebuilt.
3. Build again via `FastBuild` using `-cache` the same `Clang` compiler

The expected result is of course that the build succeeds. In our case, we get `Clang` errors:
```
31>WARNING: C:\dev\Build\Output\Android-ARM-Development\SomeProject\SomeOjbect.obj
fatal error: file 'C:\dev\SomeOtherProject\GlobalStuff.h' has been modified since the precompiled header 'C:\dev/Build/Output/Android-ARM-Development/SomeProject/stdafx.pch' was built
note: please rebuild precompiled header 'C:\dev/Build/Output/Android-ARM-Development/SomeProject/stdafx.pch'
1 error generated.
31>Failed to build Object (error 0x1) 'C:\dev\Build\Output\Android-ARM-Development\SomeProject\Unity1.obj'
```

This does seem consistent with needing to provide `-fno-pch-timestamp` to Clang to make it work properly with `FastBuild -cache` -- but, I'm still skeptical that all this time could have passed with nobody but us encountering this bug?

Thanks in advance for any assistance/ideas you might have, and I'm sorry in advance if this ends up being a stupid mistake on our part!

Layla

P.S. @ffulin my apologies that I haven't replied on other threads yet -- I've been swamped but mean to get back to those soon.

**Comment by missmah** [comment]
For reference, here's our (hopefully) temporary solution to the problem:

```
// -Xclang -fno-pch-timestamp allows Clang to use a PCH even if the header
// files that comprise it have changed timestamps since the PCH was made.
// This is needed when running fastbuild with -cache or -cacheread,
// as FastBuild cache entries which haven't changed contents remiain valid,
// even though their file timestamps may have changed (e.g. via touch).
// Easy repro:
// Do a -cache build, modify a PluginPrivate.cpp and touch InsightfulMacros.h, then build again.
// Hopefully we're just doing something wrong. We've opened a ticket with fastbuild about this:
// https://github.com/fastbuild/fastbuild/issues/346
.ClangPCHTimestampFix = ' -Xclang -fno-pch-timestamp'
```

And then for each `Clang` compiler:
```
.PCHOptions				= .CompilerOptions
                                                + .ClangPCHTimestampFix
                                                + ... // Of course there's generally more here...
```

**Comment by ffulin** [comment]
I don't think you are doing anything wrong - I think that flag is needed when using caching.

I've never managed to get a performance boost in real-world situations using clang's precompiled headers. Unity also gives you all the benefits of precompiled headers for everything but the largest of libraries (and is more efficient in some cases). With that in mind, I don't tend to use precompiled headers with Clang or gcc. I think because of this, this isn't a problem I've seen before. The PCH test is also too simple to catch this case.

With all that in mind, I think we should consider this a bug.

It seems like we should fix this by having FASTBuild automatically add the -f-no-pch-timestamp option when compiling precompiled headers using Clang. I think we can also extend the tests to cover this case.

What do you think?

**Comment by missmah** [comment]
@ffulin I think you meant to say add the `-f-no-pch-timestamp` option when using `-cachewrite`, `-cacheread` or `-cache` :) We'd also have to add the `-Xclang` before it as well. I do think this is probably safe and in the spirit of everything else fastbuild is doing (e.g. running a preprocessing step where it adds flags to flatten the headers).

You may also be correct that we see little gain with the combination of `Unity` + `PCH` - I haven't measured the difference when removing `PCH` since enabling `Unity` yet, but it sounds like I should do that. :-)

**Comment by ffulin** [comment]
I actually think we should add the necessary options for the pch in all cases, unless you know of a specific reason not to do that?

While I try to mess with the command line as little as possible, in this case I think it would be preferable for the pch to behave the same whether you are using the cache or not.

**Comment by craigscott-crascit** [comment]
> That flag (-fno-pch-timestamp) doesn't seem to exist in the clang options.

The `-fno-pch-timestamp` option is for the clang compiler, not the clang front end. It won't show if you just do `clang --help`, but it will if you do `clang -cc1 --help`. As @missmah mentioned, you have to specify it as `-Xclang -fno-pch-timestamp` on your compiler command line.
